Integrate Google AdSense for Search and Adsense for Shopping (on the Custom Search Ads protocol) with AMP

megankj opened the pull request:
This change allows publishers to request Google AdSense for Search and AdSense for Shopping on the Custom Search Ads protocol.  We introduce a new AMP ad type, called "csa". 

#2659

megankj commented:
@lannka Can you take a look at this?

megankj commented:
@jasti FYI

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+ * @param {Object} container HTML element of the CSA container
+ * @param {number} height Height to resize, in pixels
+ */
+function resizeCsa(container, height) {
```
move private method to the bottom

megankj replied:
done

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+ * @param {*} backfillPageOptions AFS page options (if necessary)
+ * @param {*} backfillAdblockOptions AFS ad unit options (if necessary)
+ * @return {function(string,boolean)} The callback function
+ */
```
add annotation `@visibleForTesting` if it's public only for testing purpose.

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+ * This is needed for an iOS bug found in versions 10.0.1 and below that
+ * doesn't properly reflow the iframe upon orientation change.
+ */
+window.addEventListener('orientationchange', function() {
```
is it necessary to have this call outside your main function?

megankj commented:
@lannka Any other feedback?

jridgewell commented:
ping @lannka

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,349 @@
+   * This is needed for an iOS bug found in versions 10.0.1 and below that
+   * doesn't properly reflow the iframe upon orientation change.
+   */
+  window.addEventListener('orientationchange', function() {
```
use arrow function

megankj replied:
done

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,349 @@
+    // Save the height of the container before the event listener triggers
+    const oldHeight = document.getElementById('csacontainer').style.height;
+
+    setTimeout(function() {
```
same here

megankj replied:
done

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,349 @@
+                          'ampSlotIndex']);
+
+  // Add the ad container to the document
+  const d = global.document.createElement('div');
```
can you give it a better name, like `containerDiv`?

megankj replied:
done

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,349 @@
+        const overflow = document.getElementById('overflow');
+        if (overflow) {
+          overflow.onclick = function() {
+            window.context.requestResize('auto', newHeight);
```
please replace all the usage of `window` to `global`.

lannka replied:
the API does not accept 'auto'. If you only want to change the height, just pass `undefined`

megankj replied:
done

megankj replied:
done

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,349 @@
+        }
+
+        // Resize the container to the correct height
+        window.context.requestResize('auto', newHeight);
```
same here

megankj replied:
done

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,349 @@
+  // When the overflow element is clicked, resize the AMP iframe
+  // to what we tried to resize before
+  overflow.onclick = function() {
+    window.context.requestResize('auto', fullH);
```
same here

megankj replied:
done

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,349 @@
+ * @param {Node} container HTML element of the CSA container
+ * @param {number} containerH Height to resize the CSA container, in pixels
+ */
+function createOverflow(overflowH, fullH, container, containerH) {
```
nit: please swap the position of `createOverFlow` and `resizeCsa`. Easier for reader to follow the logic.

megankj replied:
done

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,349 @@
+      global._googCsa('plas', afshPageOptions, afshAdblockOptions);
+    }
+  });
+
```
remove

megankj replied:
Why do you want me to remove this?  This is an ad call.

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,349 @@
+ * @return {function(string,boolean)} The callback function
+ * @visibleForTesting
+ */
+export function generateCallback(backfillPageOptions, backfillAdblockOptions) {
```
you will want to pass in the global object

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,349 @@
+
+/**
+ * Create a callback function to resize the iframe and/or request backfill
+ * @param {*} backfillPageOptions AFS page options (if necessary)
```
{?Object}

megankj replied:
If I annotate is as anything other than *, gulp check-types complains
```
found   : (*|undefined)
required: Object
    tryParseJson(data['afshPageOptions'])
```

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,349 @@
+   * @param {string} containerName
+   * @param {boolean} adsLoaded
+   */
+  const resizeIframe = function(containerName, adsLoaded) {
```
instead of return a function, we typically do:

```
afshAdblockOptions['adLoadedCallback'] = resizeIframe.bind(null, afsPageOptions, afsAdblockOptions);

function resizeIframe(backfillPageOptions, backfillAdblockOptions, containerName, adsLoaded) {
  ...
}
```

megankj replied:
done

megankj commented on ads/google/csa.js:
```diff
@@ -0,0 +1,349 @@
+  // Get parent width in case we want to override
+  const width = global.document.body.clientWidth;
+
+  validateData(data, [], ['afshPageOptions',
```
done

megankj commented:
I have deployed the latest code to Heroku as well.  Here are several demos that demonstrate various use cases.  These examples aren't in the pull request (because 8 examples is a little excessive) but we plan to document these use cases in resources that will be available to partners.

http://ampcsa.herokuapp.com/examples/afs.amp.min.html
http://ampcsa.herokuapp.com/examples/afstop.amp.min.html
http://ampcsa.herokuapp.com/examples/afstopfallback.amp.min.html
http://ampcsa.herokuapp.com/examples/afsh.amp.min.html
http://ampcsa.herokuapp.com/examples/afshbackfill.amp.min.html
http://ampcsa.herokuapp.com/examples/afshtop.amp.min.html
http://ampcsa.herokuapp.com/examples/afsafsh.amp.min.html

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+    // If the height of the container is larger than the height of the
+    // initially requested AMP tag, add the overflow element
+    if (height > currentAmpHeight) {
+
```
nitpick: To save space, please remove blank line after the `if` statement.

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+
+  // Add overflow to global.document and resize container as necessary
+  global.document.body.appendChild(overflow);
+  resizeCsa(container, containerH);
```
Probably `resizeContainer` is a clearer name?

megankj replied:
Well it resizes both the CSA container and the CSA iframe inside of it.

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+      _googCsa('ads', backfillPageOptions, backfillAdblockOptions);
+
+    } else {
+
```
remove blank line

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+
+    // Save the height of the container before the event listener triggers
+    const oldHeight =
+      global.document.getElementById('csacontainer').style.height;
```
indent 4 spaces for line break.

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+      /*eslint-enable no-unused-vars*/
+
+      // Capture new height
+      const container = global.document.getElementById('csacontainer');
```
you're still in the same closure, you can still access the `csaContainerDiv`. Nit: I think we can just name it `containerDiv`, given the context it's clear what container div it is.

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+      global._googCsa('plas', afshPageOptions, afshAdblockOptions);
+
+    } else if (data['afsPageOptions'] != null
+      && data['afshPageOptions'] != null) {
```
indentation

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+
+      // Create a callback with the AFS options
+      afshAdblockOptions['adLoadedCallback'] =
+        resizeIframe.bind(null, global, afsPageOptions, afsAdblockOptions);
```
indentation

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+    const container = global.document.getElementById(containerName);
+    const height = container.offsetHeight;
+    currentAmpHeight =
+        global.context.initialIntersection.boundingClientRect.height;
```
are you able to use `global.innerHeight`

megankj replied:
I had no idea that existed! I had to create currentAmpHeight to track the height of the AMP ad.  I will replace all instances of currentAmpHeight with global.innerHeight.  Can we get this documented somewhere on the ads README so others know it is available to use?

megankj replied:
Actually, global.innerHeight doesn't work because I need the height of the AMP container at the time of requesting the ads before the iframe is resized.  Different browsers have different behaviors (some bust the iframe, some don't).  Changing to global.innerHeight broke the overflow implementation on iOS.

lannka commented on build-system/dep-check-config.js:
```diff
@@ -98,6 +98,8 @@ exports.rules = [
       'ads/alp/handler.js->src/dom.js',
       'ads/alp/handler.js->src/config.js',
+      // CSA needs to depend on src files
+      'ads/google/csa.js->src/json.js',
```
let's just allow all `ads/**` to depend on `json.js`.

lannka commented on examples/csa.amp.html:
```diff
@@ -0,0 +1,54 @@
+    </amp-ad>
+  </div>
+</body>
+</html>
```
insert blank line at the bottom of file

lannka commented on test/functional/ads/test-csa.js:
```diff
@@ -0,0 +1,348 @@
+            height: 0,
+          },
+        },
+        requestResize: function() {},
```
can be simplified to

`requestResize() {},`

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+    {container: containerId});
+
+  /* Time in ms to wait before executing orientation change function logic */
+  const orientationChangeTimeout = 250;
```
you can just inline this number, and move comment there

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+
+    }, orientationChangeTimeout);
+
+  }, false);
```
`false` is not needed, it's the default

lannka commented on examples/csa.amp.html:
```diff
@@ -0,0 +1,54 @@
+  <link rel="canonical" href="http://nonblocking.io/" >
+  <meta name="viewport" content="width=device-width,minimum-scale=1,initial-scale=1">
+  <style amp-custom>
+    amp-ad {
```
remove empty CSS class

lannka commented on test/functional/ads/test-csa.js:
```diff
@@ -0,0 +1,348 @@
+import {createIframePromise} from '../../../testing/iframe';
+import {csa, resizeIframe} from '../../../ads/google/csa';
+import * as sinon from 'sinon';
+import * as p3p from '../../../3p/3p';
```
`as _3p`

lannka commented on test/functional/ads/test-csa.js:
```diff
@@ -0,0 +1,348 @@
+  let sandbox;
+  let win;
+
+  function getAds(type) {
```
you can probably turn the whole thing to a  constant map

megankj replied:
Won't a map return an array though?  I need an object.

lannka commented on test/functional/ads/test-csa.js:
```diff
@@ -0,0 +1,348 @@
+
+  beforeEach(() => {
+    sandbox = sinon.sandbox.create();
+    return createIframePromise(true).then(iframe => {
```
please try to use `describes.fakeWin` instead.

lannka commented on test/functional/ads/test-csa.js:
```diff
@@ -0,0 +1,348 @@
+  });
+
+  describe('inputs', () => {
+
```
remove blank line

lannka commented on test/functional/ads/test-csa.js:
```diff
@@ -0,0 +1,348 @@
+    it('should create a csa container', () => {
+      csa(win, getAds('afs'));
+      const container = win.document.getElementById('csacontainer');
+      expect(container).not.to.equal(null);
```
expect(container).not.to.be.null

megankj commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+  // Parse AFSh page options
+  const afshPageOptions = Object.assign(
+    Object(tryParseJson(data['afshPageOptions'])),
+    {source: 'amp', referer: global.context.referrer});
```
I want to first parse the what the partner sends in and then add these parameters because the second parameter override the first parameter.  If the partner passes in "source" or "referer", we will lose the values we're setting manually.  It's probably a rare case but doing it in this order ensures these values are set correctly.

lannka commented on test/functional/ads/test-csa.js:
```diff
@@ -0,0 +1,348 @@
+      resizeIframe(win, {}, {}, 'csacontainer', false);
+
+      // Should not tell AMP we have no ads
+      expect(noAdsSpy.called).to.equal(false);
```
expect(noadsSpy).not.to.be.called

megankj commented on ads/google/csa.js:
```diff
@@ -0,0 +1,327 @@
+
+      // Force DOM reflow and repaint
+      /*eslint-disable no-unused-vars*/
+      const ignore = global.document.body.offsetHeight;
```
You would get a warning from gulp check-types

```
ads/google/csa.js:88: WARNING - Suspicious code. The result of the 'getprop' operator is not being used.
      global.document.body.offsetHeight;
```

megankj replied:
Also this section was causing me compiler errors when I tried to deploy to heroku (#5351) so this code was added to fix that.

megankj commented:
There is some weirdness I can't explain with iOS9 (9.3.5) that shows the overflow element on top of the adblock.  When I inspect it, the heights for the AMP ad, AMP iframe, CSA div, and CSA iframe are all set correctly.  I tried setting overflow-y: 'hidden' in the code and that didn't fix the problem either.  If I scroll down and back up, the problem fixes itself.  If I play around with the CSS properties (and don't actually change anything, just set it back to what it was before) it fixes itself.

Not sure what's going on here but I'll investigate more on Monday.

megankj commented:
On iOS9, when a resize is requested _sometimes_ the CSA iframe busts and then returns to it's full height even though we have set the height via CSS.  I was able to stop this behavior by setting the iframe width to "100%", but you can still see a flicker of the ad expanding and contracting.  For some reason this only happens some of the time on iOS9 and not at all on iOS10 or Android.  Perhaps this is related to the iframe reflow issue that required me to add an event listener for orientation change.

One possible way to stop the flicker is to make the overflow background a solid color and not transparent, but that locks partners into a particular styling which we probably don't want to do.

Test page:
http://ampcsa.herokuapp.com/examples/afstop.amp.min.html

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,264 @@
+    });
+
+    // If resize was denied, show overflow element and resize container
+    global.context.onResizeDenied(function(requestedHeight) {
```
will there be a problem when `resizeIframe()` get called multiple times? Seems we're going to register multiple listener here, which could create multiple overflow elements?

megankj replied:
Resize iframe is only called multiple times if the partner wants to make a new request for (AFS) ads if the first (AFSh) request didn't return anything.  In this case, we won't have made an overflow or created any listeners because adsLoaded in resizeIframe will be false.

The resizeDenied and resizeSuccess functions check for existence of overflow before creating a new one and/or resizing.

lannka replied:
Even though, I think logically it's better to move this logic out of the resize handler, to make sure we only register once.

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,264 @@
+  const pageOptions = {source: 'amp', referer: global.context.referrer};
+  const adblockOptions = {container: containerId};
+  const afshPageOptions = Object.assign(
+      Object(tryParseJson(data['afshPageOptions'])), pageOptions);
```
did you try swap the argument position?  That will help you get rid of `Object()`.

```
Object.assign(pageOptions, tryParseJson(data['afshPageOptions']));
```

lannka replied:
Ah, never mind. that will mutate `pageOptions`.

megankj replied:
Switching the order does not get rid of the check-types error.  Furthermore, if a partner passes in "source" or "referrer" they would end up overriding what we set if we do it in this order.

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,264 @@
+  global.addEventListener('orientationchange', () => {
+    // Save the height of the container before the event listener triggers
+    const oldHeight = containerDiv.style.height;
+    setTimeout(() => {
```
global.setTimeout

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,264 @@
+   */
+  global.addEventListener('orientationchange', () => {
+    // Save the height of the container before the event listener triggers
+    const oldHeight = containerDiv.style.height;
```
could you extract out the whole listener to a function?

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,264 @@
+
+  // Only call for ads once the script has loaded
+  loadScript(global, 'https://www.google.com/adsense/search/ads.js', () => {
+    if (data['afsPageOptions'] != null && data['afshPageOptions'] == null) {
```
could you extract out the load callback to a function?

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,305 @@
+  }
+
+  // Event listener needed for iOS9 bug
+  global.addEventListener('orientationchange', () => {
```
global.addEventListener('orientationchange', orientationChangeHandler.bind(null, global, containerDiv));

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,305 @@
+
+  // Only call for ads once the script has loaded
+  loadScript(global, 'https://www.google.com/adsense/search/ads.js', () => {
+    requestCsaAds(global, data, afsPage, afsAd, afshPage, afshAd);
```
use bind here as well

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,305 @@
+let currentAmpHeight = null;
+
+// Enum to hold different ad types
+const ADTYPE = {
```
`/** @enum {string} */`

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,305 @@
+// Enum to hold different ad types
+const ADTYPE = {
+  UNSUPPORTED: 0,
+  AFS: 1,
```
can you add some description about the enum value?

lannka commented on test/functional/ads/test-csa.js:
```diff
@@ -0,0 +1,307 @@
+import * as sinon from 'sinon';
+import * as _3p from '../../../3p/3p';
+
+describe('amp-ad-csa-impl', () => {
```
seems this `describe` wrapper can be omitted?

lannka commented on ads/google/csa.js:
```diff
@@ -19,14 +19,25 @@ import {tryParseJson} from '../../src/json.js';
+
+/**
+ * Enum for different AdSense Products
+ * UNSUPPORTED: Value if we can't determine which product to request
```
nitpick, you can inline descriptions.
```
/** Value if we can't determine which product to request */
UNSUPPORTED: 0,
```

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,324 @@
+      Object(tryParseJson(data['afsAdblockOptions'])), adblockOptions);
+
+  // Special case for AFSh when "auto" is the requested width
+  if (afshAd != null && afshAd['width'] == 'auto') {
```
seems `afshAd` is guaranteed not null

megankj commented:
I addressed your comments and tried to remove unnecessary code to make it more compact (and in some cases fit on one line).  Let me know if I should reorder the helper functions or do anything else to clean it up.

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,324 @@
+ * @param {?Object} afsP The parsed AFS page options object
+ * @param {?Object} afsA The parsed AFS adblock options object
+ * @param {?Object} afshP The parsed AFSh page options object
+ * @param {?Object} afshA The parsed AFSh adblock options object
```
seems none of the arguments can be null

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,324 @@
+ */
+function requestCsaAds(global, data, afsP, afsA, afshP, afshA) {
+  const type = getAdType(data);
+  if (type == ADTYPE.AFS) {
```
is `switch` more appropriate here?

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,324 @@
+ * AFSHBACKFILL: AdSense for Shopping, backfilled with AdSense for Search
+ * @enum {number}
+ */
+const ADTYPE = {
```
AD_TYPE

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,324 @@
+  UNSUPPORTED: 0,
+  AFS: 1,
+  AFSH: 2,
+  AFSHBACKFILL: 3,
```
AFSH_BACKFILL

lannka commented on test/functional/ads/test-csa.js:
```diff
@@ -0,0 +1,311 @@
+ * AFSHBACKFILL: AdSense for Shopping, backfilled with AdSense for Search
+ * @enum {number}
+ */
+const ADTYPE = {
```
can you just import this enum?

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,332 @@
+ * @param {boolean} ads If ads were returned or not
+ * @visibleForTesting
+ */
+export function resizeIframe(global, bkflPage, bkflAd, conName, ads) {
```
shall we just split this function into 2?
```
function resizeIframe(global, containerName, hasAd) {}
function backfill(global, backfillPage, backfillAd) {}
```

it gives not just shorter argument list, but also avoided the recursive call of resizeIframe.

Then the above listener code becomes:
```
afshA['adLoadedCallback'] = (containerName, hasAd) => {
  if (hasAd) {
    resizeIframe(global, containerName, hasAd);
  } else {
     backfill(global, afsP, afsA)
  }
} 
```

lannka commented on ads/google/csa.js:
```diff
@@ -0,0 +1,361 @@
+ * @param {?Object} page Backfill AFS page options (if necessary)
+ * @param {?Object} adblock Backfill AFS ad unit options (if necessary)
+ */
+function backfillAds(global, page, adblock) {
```
this little function is only used once, you can inline.
